// @iota/transaction-converter/out/transaction-converter/src/index.js
"use strict";
/** @module transaction-converter */
exports.__esModule = true;
var converter_1 = require("@iota/converter");
var pad_1 = require("@iota/pad");
var transaction_1 = require("@iota/transaction");
var errors = require("../../errors");
var guards_1 = require("../../guards");
require("../../typed-array");
var types_1 = require("../../types");
/**
 * Converts a transaction object or a list of those into transaction trytes.
 *
 * @method asTransactionTrytes
 *
 * @param {Transaction | Transaction[]} transactions - Transaction object(s)
 *
 * @return {Trytes | Trytes[]} Transaction trytes
 */
function asTransactionTrytes(transactions) {
    var txTrytes = types_1.asArray(transactions).map(function (transaction) {
        return [
            transaction.signatureMessageFragment,
            transaction.address,
            converter_1.tritsToTrytes(pad_1.padTrits(81)(converter_1.trytesToTrits(transaction.value))),
            pad_1.padTrytes(27)(transaction.obsoleteTag),
            converter_1.tritsToTrytes(pad_1.padTrits(27)(converter_1.trytesToTrits(transaction.timestamp))),
            converter_1.tritsToTrytes(pad_1.padTrits(27)(converter_1.trytesToTrits(transaction.currentIndex))),
            converter_1.tritsToTrytes(pad_1.padTrits(27)(converter_1.trytesToTrits(transaction.lastIndex))),
            transaction.bundle,
            transaction.trunkTransaction,
            transaction.branchTransaction,
            pad_1.padTrytes(27)(transaction.tag || transaction.obsoleteTag),
            converter_1.tritsToTrytes(pad_1.padTrits(27)(converter_1.trytesToTrits(transaction.attachmentTimestamp))),
            converter_1.tritsToTrytes(pad_1.padTrits(27)(converter_1.trytesToTrits(transaction.attachmentTimestampLowerBound))),
            converter_1.tritsToTrytes(pad_1.padTrits(27)(converter_1.trytesToTrits(transaction.attachmentTimestampUpperBound))),
            transaction.nonce,
        ].join('');
    });
    return Array.isArray(transactions) ? txTrytes : txTrytes[0];
}
exports.asTransactionTrytes = asTransactionTrytes;
/**
 * Converts transaction trytes of 2673 trytes into a transaction object.
 *
 * @method asTransactionObject
 *
 * @param {Trytes} trytes - Transaction trytes
 *
 * @return {Transaction} Transaction object
 */
exports.asTransactionObject = function (trytes, hash) {
    if (!guards_1.isTrytesOfExactLength(trytes, 2673)) {
        throw new Error(errors.INVALID_TRYTES);
    }
    for (var i = 2279; i < 2295; i++) {
        if (trytes.charAt(i) !== '9') {
            throw new Error(errors.INVALID_TRYTES);
        }
    }
    var trits = converter_1.trytesToTrits(trytes);
    return {
        hash: hash || converter_1.tritsToTrytes(transaction_1.transactionHash(trits)),
        signatureMessageFragment: trytes.slice(0, 2187),
        address: trytes.slice(2187, 2268),
        value: converter_1.value(trits.slice(6804, 6837)),
        obsoleteTag: trytes.slice(2295, 2322),
        timestamp: converter_1.value(trits.slice(6966, 6993)),
        currentIndex: converter_1.value(trits.slice(6993, 7020)),
        lastIndex: converter_1.value(trits.slice(7020, 7047)),
        bundle: trytes.slice(2349, 2430),
        trunkTransaction: trytes.slice(2430, 2511),
        branchTransaction: trytes.slice(2511, 2592),
        tag: trytes.slice(2592, 2619),
        attachmentTimestamp: converter_1.value(trits.slice(7857, 7884)),
        attachmentTimestampLowerBound: converter_1.value(trits.slice(7884, 7911)),
        attachmentTimestampUpperBound: converter_1.value(trits.slice(7911, 7938)),
        nonce: trytes.slice(2646, 2673)
    };
};
/**
 * Converts a list of transaction trytes into list of transaction objects.
 * Accepts a list of hashes and returns a mapper. In cases hashes are given,
 * the mapper function map them to converted objects.
 *
 * @method asTransactionObjects
 *
 * @param {Hash[]} [hashes] - Optional list of known hashes.
 * Known hashes are directly mapped to transaction objects,
 * otherwise all hashes are being recalculated.
 *
 * @return {Function} {@link #module_transaction.transactionObjectsMapper `transactionObjectsMapper`}
 */
exports.asTransactionObjects = function (hashes) {
    /**
     * Maps the list of given hashes to a list of converted transaction objects.
     *
     * @method transactionObjectsMapper
     *
     * @param {Trytes[]} trytes - List of transaction trytes to convert
     *
     * @return {Transaction[]} List of transaction objects with hashes
     */
    return function transactionObjectsMapper(trytes) {
        return trytes.map(function (tryteString, i) { return exports.asTransactionObject(tryteString, hashes[i]); });
    };
};
exports.asFinalTransactionTrytes = function (transactions) {
    return asTransactionTrytes(transactions).slice().reverse();
};
exports.transactionObject = function (trytes) {
    /* tslint:disable-next-line:no-console */
    console.warn('`transactionObject` has been renamed to `asTransactionObject`');
    return exports.asTransactionObject(trytes);
};
exports.transactionTrytes = function (transaction) {
    /* tslint:disable-next-line:no-console */
    console.warn('`transactionTrytes` has been renamed to `asTransactionTrytes`');
    return asTransactionTrytes(transaction);
};
