// @iota/transaction/out/transaction/src/index.js
"use strict";
exports.__esModule = true;
var transaction_1 = require("./transaction");
exports.address = transaction_1.address;
exports.attachmentTimestamp = transaction_1.attachmentTimestamp;
exports.attachmentTimestampLowerBound = transaction_1.attachmentTimestampLowerBound;
exports.attachmentTimestampUpperBound = transaction_1.attachmentTimestampUpperBound;
exports.branchTransaction = transaction_1.branchTransaction;
exports.bundle = transaction_1.bundle;
exports.currentIndex = transaction_1.currentIndex;
exports.lastIndex = transaction_1.lastIndex;
exports.obsoleteTag = transaction_1.obsoleteTag;
exports.isAttached = transaction_1.isAttached;
exports.isHead = transaction_1.isHead;
exports.isMultipleOfTransactionLength = transaction_1.isMultipleOfTransactionLength;
exports.issuanceTimestamp = transaction_1.issuanceTimestamp;
exports.isTail = transaction_1.isTail;
exports.isTransaction = transaction_1.isTransaction;
exports.signatureOrMessage = transaction_1.signatureOrMessage;
exports.tag = transaction_1.tag;
exports.transactionEssence = transaction_1.transactionEssence;
exports.transactionHash = transaction_1.transactionHash;
exports.transactionNonce = transaction_1.transactionNonce;
exports.trunkTransaction = transaction_1.trunkTransaction;
exports.value = transaction_1.value;
exports.SIGNATURE_OR_MESSAGE_OFFSET = transaction_1.SIGNATURE_OR_MESSAGE_OFFSET;
exports.SIGNATURE_OR_MESSAGE_LENGTH = transaction_1.SIGNATURE_OR_MESSAGE_LENGTH;
exports.ADDRESS_OFFSET = transaction_1.ADDRESS_OFFSET;
exports.ADDRESS_LENGTH = transaction_1.ADDRESS_LENGTH;
exports.VALUE_OFFSET = transaction_1.VALUE_OFFSET;
exports.VALUE_LENGTH = transaction_1.VALUE_LENGTH;
exports.OBSOLETE_TAG_OFFSET = transaction_1.OBSOLETE_TAG_OFFSET;
exports.OBSOLETE_TAG_LENGTH = transaction_1.OBSOLETE_TAG_LENGTH;
exports.ISSUANCE_TIMESTAMP_OFFSET = transaction_1.ISSUANCE_TIMESTAMP_OFFSET;
exports.ISSUANCE_TIMESTAMP_LENGTH = transaction_1.ISSUANCE_TIMESTAMP_LENGTH;
exports.CURRENT_INDEX_OFFSET = transaction_1.CURRENT_INDEX_OFFSET;
exports.CURRENT_INDEX_LENGTH = transaction_1.CURRENT_INDEX_LENGTH;
exports.LAST_INDEX_OFFSET = transaction_1.LAST_INDEX_OFFSET;
exports.LAST_INDEX_LENGTH = transaction_1.LAST_INDEX_LENGTH;
exports.BUNDLE_OFFSET = transaction_1.BUNDLE_OFFSET;
exports.BUNDLE_LENGTH = transaction_1.BUNDLE_LENGTH;
exports.TRUNK_TRANSACTION_OFFSET = transaction_1.TRUNK_TRANSACTION_OFFSET;
exports.TRUNK_TRANSACTION_LENGTH = transaction_1.TRUNK_TRANSACTION_LENGTH;
exports.BRANCH_TRANSACTION_OFFSET = transaction_1.BRANCH_TRANSACTION_OFFSET;
exports.BRANCH_TRANSACTION_LENGTH = transaction_1.BRANCH_TRANSACTION_LENGTH;
exports.TAG_OFFSET = transaction_1.TAG_OFFSET;
exports.TAG_LENGTH = transaction_1.TAG_LENGTH;
exports.ATTACHMENT_TIMESTAMP_OFFSET = transaction_1.ATTACHMENT_TIMESTAMP_OFFSET;
exports.ATTACHMENT_TIMESTAMP_LENGTH = transaction_1.ATTACHMENT_TIMESTAMP_LENGTH;
exports.ATTACHMENT_TIMESTAMP_LOWER_BOUND_OFFSET = transaction_1.ATTACHMENT_TIMESTAMP_LOWER_BOUND_OFFSET;
exports.ATTACHMENT_TIMESTAMP_LOWER_BOUND_LENGTH = transaction_1.ATTACHMENT_TIMESTAMP_LOWER_BOUND_LENGTH;
exports.ATTACHMENT_TIMESTAMP_UPPER_BOUND_OFFSET = transaction_1.ATTACHMENT_TIMESTAMP_UPPER_BOUND_OFFSET;
exports.ATTACHMENT_TIMESTAMP_UPPER_BOUND_LENGTH = transaction_1.ATTACHMENT_TIMESTAMP_UPPER_BOUND_LENGTH;
exports.TRANSACTION_NONCE_OFFSET = transaction_1.TRANSACTION_NONCE_OFFSET;
exports.TRANSACTION_NONCE_LENGTH = transaction_1.TRANSACTION_NONCE_LENGTH;
exports.TRANSACTION_ESSENCE_OFFSET = transaction_1.TRANSACTION_ESSENCE_OFFSET;
exports.TRANSACTION_ESSENCE_LENGTH = transaction_1.TRANSACTION_ESSENCE_LENGTH;
exports.TRANSACTION_LENGTH = transaction_1.TRANSACTION_LENGTH;
exports.TRANSACTION_HASH_LENGTH = transaction_1.TRANSACTION_HASH_LENGTH;
