// @iota/core/out/core/test/integration/getAccountData.test.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
exports.__esModule = true;
var http_client_1 = require("@iota/http-client");
var samples_1 = require("@iota/samples");
var ava_1 = require("ava");
var errors_1 = require("../../../errors");
var src_1 = require("../../src");
require("./nocks/findTransactions");
var getBalances_1 = require("./nocks/getBalances");
require("./nocks/getInclusionStates");
require("./nocks/getNodeInfo");
require("./nocks/getTrytes");
require("./nocks/wereAddressesSpentFrom");
var getAccountData = src_1.createGetAccountData(http_client_1.createHttpClient(), 'lib');
var seed = 'SEED';
var balances = getBalances_1.balancesResponse.balances;
var accountAddresses = [
    getBalances_1.getBalancesCommand.addresses[0],
    getBalances_1.getBalancesCommand.addresses[1],
    getBalances_1.getBalancesCommand.addresses[2],
];
var accountData = {
    addresses: accountAddresses,
    inputs: [
        {
            address: getBalances_1.getBalancesCommand.addresses[2],
            balance: balances[2],
            keyIndex: 2,
            security: 2
        },
    ],
    latestAddress: getBalances_1.getBalancesCommand.addresses[2],
    transfers: samples_1.transfers,
    transactions: samples_1.transfers.reduce(function (acc, bundle) {
        return acc.concat(bundle
            .filter(function (_a) {
            var address = _a.address;
            return accountAddresses.indexOf(address) > -1;
        })
            .map(function (transaction) { return transaction.hash; }));
    }, []),
    balance: 1
};
ava_1["default"]('getAccountData() resolves to correct account data', function (t) { return __awaiter(_this, void 0, void 0, function () {
    var _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _b = (_a = t).deepEqual;
                return [4 /*yield*/, getAccountData(seed, { start: 0 })];
            case 1:
                _b.apply(_a, [_c.sent(),
                    accountData,
                    'getAccountData() should resolve to correct account data']);
                return [2 /*return*/];
        }
    });
}); });
ava_1["default"]('getAccountData() rejects with correct errors for invalid inputs', function (t) {
    var invalidSeed = 'asdasDSFDAFD';
    var invalidStartEndOptions = {
        start: 10,
        end: 9
    };
    t.is(t.throws(function () { return getAccountData(invalidSeed, { start: 0 }); }, Error).message, errors_1.INVALID_SEED + ": " + invalidSeed, 'getAccountData() should throw correct error for invalid seed');
    t.is(t.throws(function () { return getAccountData(seed, invalidStartEndOptions); }, Error).message, errors_1.INVALID_START_END_OPTIONS + ": " + invalidStartEndOptions, 'getAccountData() should throw correct error for invalid start & end options');
});
ava_1["default"]('getAccountData() with start > 0 resolves to correct account data.', function (t) { return __awaiter(_this, void 0, void 0, function () {
    var expected, _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                expected = __assign({}, accountData, { addresses: accountData.addresses.slice(1, 3), balance: getBalances_1.balancesResponse.balances.slice(1, 3).reduce(function (acc, balance) { return acc + balance; }, 0) });
                _b = (_a = t).deepEqual;
                return [4 /*yield*/, getAccountData(seed, { start: 1 })];
            case 1:
                _b.apply(_a, [_c.sent(),
                    expected,
                    'getAccountData() with start > 0 should resolve to correct account data']);
                return [2 /*return*/];
        }
    });
}); });
ava_1["default"].cb('getAccountData() invokes callback', function (t) {
    getAccountData(seed, { start: 0 }, t.end);
});
ava_1["default"].cb('getAccountData() passes correct arguments to callback', function (t) {
    getAccountData(seed, { start: 0 }, function (err, res) {
        t.is(err, null, 'getAccountData() should pass null as first argument in callback for successuful requests');
        t.deepEqual(res, accountData, 'getAccountData() should pass the correct response as second argument in callback');
        t.end();
    });
});
