// @iota/core/out/core/test/integration/nocks/checkConsistency.js
"use strict";
exports.__esModule = true;
var nock = require("nock");
var types_1 = require("../../../../types");
var headers_1 = require("./headers");
exports.checkConsistencyCommand = {
    command: types_1.IRICommand.CHECK_CONSISTENCY,
    tails: ['A'.repeat(81), 'B'.repeat(81)]
};
exports.checkConsistencyResponse = {
    state: true,
    info: ''
};
exports.checkConsistencyNock = nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', exports.checkConsistencyCommand)
    .reply(200, exports.checkConsistencyResponse);
exports.checkConsistencyWithInfoCommand = {
    command: types_1.IRICommand.CHECK_CONSISTENCY,
    tails: ['A'.repeat(81), 'B'.repeat(81), 'C'.repeat(81)]
};
exports.checkConsistencyWithInfoResponse = {
    state: false,
    info: 'test response'
};
exports.checkConsistencyWithInfoNock = nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', exports.checkConsistencyWithInfoCommand)
    .reply(200, exports.checkConsistencyWithInfoResponse);
