// @iota/core/out/core/test/integration/nocks/getTrytes.js
"use strict";
exports.__esModule = true;
var samples_1 = require("@iota/samples");
var nock = require("nock");
var types_1 = require("../../../../types");
var headers_1 = require("./headers");
exports.getTrytesCommand = {
    command: types_1.IRICommand.GET_TRYTES,
    hashes: ['A'.repeat(81), 'B'.repeat(81)]
};
exports.getTrytesResponse = {
    trytes: ['9'.repeat(2673), '9'.repeat(2673)]
};
exports.getBalancesNock = nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', exports.getTrytesCommand)
    .reply(200, exports.getTrytesResponse);
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.GET_TRYTES,
    hashes: [samples_1.bundleWithZeroValue[0].hash]
})
    .reply(200, {
    trytes: samples_1.bundleWithZeroValueTrytes
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.GET_TRYTES,
    hashes: [samples_1.bundle[0].hash]
})
    .reply(200, {
    trytes: [samples_1.bundleTrytes[0]]
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.GET_TRYTES,
    hashes: [samples_1.bundle[1].hash]
})
    .reply(200, {
    trytes: [samples_1.bundleTrytes[1]]
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.GET_TRYTES,
    hashes: [samples_1.bundle[2].hash]
})
    .reply(200, {
    trytes: [samples_1.bundleTrytes[2]]
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.GET_TRYTES,
    hashes: [samples_1.bundle[3].hash]
})
    .reply(200, {
    trytes: [samples_1.bundleTrytes[3]]
});
