// @iota/core/out/core/test/integration/nocks/getTransactionsToApprove.js
"use strict";
exports.__esModule = true;
var samples_1 = require("@iota/samples");
var nock = require("nock");
var types_1 = require("../../../../types");
var headers_1 = require("./headers");
exports.depth = 3;
exports.getTransactionsToApproveCommand = {
    command: types_1.IRICommand.GET_TRANSACTIONS_TO_APPROVE,
    depth: exports.depth
};
exports.getTransactionsToApproveWithReferenceCommand = {
    command: types_1.IRICommand.GET_TRANSACTIONS_TO_APPROVE,
    depth: exports.depth,
    reference: 'R'.repeat(81)
};
exports.getTransactionsToApproveResponse = {
    trunkTransaction: samples_1.bundle[samples_1.bundle.length - 1].trunkTransaction,
    branchTransaction: samples_1.bundle[samples_1.bundle.length - 1].branchTransaction,
    duration: 10
};
exports.getTransactionsToApproveWithReferenceResponse = exports.getTransactionsToApproveResponse;
exports.getTransactionsToApproveNock = nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', exports.getTransactionsToApproveCommand)
    .reply(200, exports.getTransactionsToApproveResponse);
exports.getTransactionsToApproveWithReferenceNock = nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', exports.getTransactionsToApproveWithReferenceCommand)
    .reply(200, exports.getTransactionsToApproveResponse);
