// @iota/core/out/core/test/integration/nocks/wereAddressesSpentFrom.js
"use strict";
exports.__esModule = true;
var nock = require("nock");
var types_1 = require("../../../../types");
var headers_1 = require("./headers");
var samples_1 = require("@iota/samples");
exports.wereAddressesSpentFromCommand = {
    command: types_1.IRICommand.WERE_ADDRESSES_SPENT_FROM,
    addresses: samples_1.addresses
};
exports.wereAddressesSpentFromResponse = {
    states: [true, false, false]
};
exports.wereAddressesSpentFromNock = nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', exports.wereAddressesSpentFromCommand)
    .reply(200, exports.wereAddressesSpentFromResponse);
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.WERE_ADDRESSES_SPENT_FROM,
    addresses: [samples_1.addresses[0]]
})
    .reply(200, {
    states: [true]
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.WERE_ADDRESSES_SPENT_FROM,
    addresses: [samples_1.addresses[1]]
})
    .reply(200, {
    states: [false]
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.WERE_ADDRESSES_SPENT_FROM,
    addresses: [samples_1.addresses[2]]
})
    .reply(200, {
    states: [false]
});
nock('http://localhost:14265', headers_1["default"])
    .persist()
    .post('/', {
    command: types_1.IRICommand.WERE_ADDRESSES_SPENT_FROM,
    addresses: [samples_1.addresses[1], samples_1.addresses[2]]
})
    .reply(200, {
    states: [false, false]
});
