// @iota/core/out/core/src/index.js
"use strict";
/** @module core */
exports.__esModule = true;
// IRI commands
var types_1 = require("../../types");
exports.IRICommand = types_1.IRICommand;
var createAddNeighbors_1 = require("./createAddNeighbors");
exports.createAddNeighbors = createAddNeighbors_1.createAddNeighbors;
var createAttachToTangle_1 = require("./createAttachToTangle");
exports.createAttachToTangle = createAttachToTangle_1.createAttachToTangle;
var createBroadcastTransactions_1 = require("./createBroadcastTransactions");
exports.createBroadcastTransactions = createBroadcastTransactions_1.createBroadcastTransactions;
var createCheckConsistency_1 = require("./createCheckConsistency");
exports.createCheckConsistency = createCheckConsistency_1.createCheckConsistency;
var createFindTransactions_1 = require("./createFindTransactions");
exports.createFindTransactions = createFindTransactions_1.createFindTransactions;
var createGetBalances_1 = require("./createGetBalances");
exports.createGetBalances = createGetBalances_1.createGetBalances;
var createGetInclusionStates_1 = require("./createGetInclusionStates");
exports.createGetInclusionStates = createGetInclusionStates_1.createGetInclusionStates;
var createGetNeighbors_1 = require("./createGetNeighbors");
exports.createGetNeighbors = createGetNeighbors_1.createGetNeighbors;
var createGetNodeInfo_1 = require("./createGetNodeInfo");
exports.createGetNodeInfo = createGetNodeInfo_1.createGetNodeInfo;
var createGetTips_1 = require("./createGetTips");
exports.createGetTips = createGetTips_1.createGetTips;
var createGetTransactionsToApprove_1 = require("./createGetTransactionsToApprove");
exports.createGetTransactionsToApprove = createGetTransactionsToApprove_1.createGetTransactionsToApprove;
var createGetTrytes_1 = require("./createGetTrytes");
exports.createGetTrytes = createGetTrytes_1.createGetTrytes;
var createInterruptAttachingToTangle_1 = require("./createInterruptAttachingToTangle");
exports.createInterruptAttachingToTangle = createInterruptAttachingToTangle_1.createInterruptAttachingToTangle;
var createRemoveNeighbors_1 = require("./createRemoveNeighbors");
exports.createRemoveNeighbors = createRemoveNeighbors_1.createRemoveNeighbors;
var createStoreTransactions_1 = require("./createStoreTransactions");
exports.createStoreTransactions = createStoreTransactions_1.createStoreTransactions;
// `wereAddressesSpentFrom` command is a temporary measure to prevent loss of funds,
// when security assumptions are ignored by developers or wallet users.
// It's being used internally by `getNewAddress()`.
// Avoid developing programs that rely on this method.
//
// export {
//     createWereAddressesSpentFrom,
//     WereAddressesSpentFromCommand,
//     WereAddressesSpentFromResponse
// } from './createWereAddressesSpentFrom'
// Wrappers
var createBroadcastBundle_1 = require("./createBroadcastBundle");
exports.createBroadcastBundle = createBroadcastBundle_1.createBroadcastBundle;
var createFindTransactionObjects_1 = require("./createFindTransactionObjects");
exports.createFindTransactionObjects = createFindTransactionObjects_1.createFindTransactionObjects;
var createGetAccountData_1 = require("./createGetAccountData");
exports.createGetAccountData = createGetAccountData_1.createGetAccountData;
var createGetBundle_1 = require("./createGetBundle");
exports.createGetBundle = createGetBundle_1.createGetBundle;
// `getBundlesFromAddress` has been deprecated because of its poor performance.
// Traversing and validating bundles gets slower as bundle instances increase.
// Use `findTransactionObjects` on `addresses` and lazily fetch the bundles when needed.
//
// export { createGetBundlesFromAddresses } from './createGetBundlesFromAddresses'
var createGetInputs_1 = require("./createGetInputs");
exports.createGetInputs = createGetInputs_1.createGetInputs;
var createGetLatestInclusion_1 = require("./createGetLatestInclusion");
exports.createGetLatestInclusion = createGetLatestInclusion_1.createGetLatestInclusion;
var createGetNewAddress_1 = require("./createGetNewAddress");
exports.createGetNewAddress = createGetNewAddress_1.createGetNewAddress;
// createGetUntilFirstUnusedAddress,
exports.createIsAddressUsed = createGetNewAddress_1.createIsAddressUsed;
var createGetTransactionObjects_1 = require("./createGetTransactionObjects");
exports.createGetTransactionObjects = createGetTransactionObjects_1.createGetTransactionObjects;
// `getTransfers` has been deprecated because of poor performance (regenerates addresses
// and calls `getBundlesFromAddresses`).
// Use `findTransactionObjects` as replacement and prefer to lazily fetch complete bundles,
// if and when required.
//
// export {
//     createGetTransfers,
//     getTransfersOptions,
//     GetTransfersOptions
// } from './createGetTransfers'
var createIsPromotable_1 = require("./createIsPromotable");
exports.isAboveMaxDepth = createIsPromotable_1.isAboveMaxDepth;
exports.createIsPromotable = createIsPromotable_1.createIsPromotable;
var createIsReattachable_1 = require("./createIsReattachable");
exports.createIsReattachable = createIsReattachable_1.createIsReattachable;
var createPromoteTransaction_1 = require("./createPromoteTransaction");
exports.createPromoteTransaction = createPromoteTransaction_1.createPromoteTransaction;
var createReplayBundle_1 = require("./createReplayBundle");
exports.createReplayBundle = createReplayBundle_1.createReplayBundle;
var createSendTrytes_1 = require("./createSendTrytes");
exports.createSendTrytes = createSendTrytes_1.createSendTrytes;
var createPrepareTransfers_1 = require("./createPrepareTransfers");
exports.createPrepareTransfers = createPrepareTransfers_1.createPrepareTransfers;
var createStoreAndBroadcast_1 = require("./createStoreAndBroadcast");
exports.createStoreAndBroadcast = createStoreAndBroadcast_1.createStoreAndBroadcast;
var createTraverseBundle_1 = require("./createTraverseBundle");
exports.createTraverseBundle = createTraverseBundle_1.createTraverseBundle;
var generateAddress_1 = require("./generateAddress");
exports.generateAddress = generateAddress_1.generateAddress;
// Errors
var errors = require("../../errors");
exports.errors = errors;
// export api factory with default provider
var composeAPI_1 = require("./composeAPI");
exports.composeAPI = composeAPI_1.composeAPI;
